panel/widgets/tree.py

```diff
@@ -0,0 +1,238 @@
+        Whether to sort nodes alphabetically.""")
+
+    show_icons = param.Boolean(default=True, doc="Whether to use icons or not.")
+
+    show_dots = param.Boolean(default=False, doc="""
+        Whether to show dots on the left as part of the tree.""")
+
+    show_stripes = param.Boolean(default=False, doc="""
+        Whether to show stripes on alternating rows.""")
+
+    value = param.List(default=[], doc="""
+        List of currently selected leaves and nodes""")
+
+    _new_nodes = param.List(default=None, doc="Children to push to tree")
+
+    _rename: ClassVar[Mapping[str, str | None]] = {
+        "icon_size": None,
+        "select_multiple": "multiple",
+        "name": None,
+        "value": "checked",
+    }
+
+    def __init__(self, **params):
+        click_handler = params.pop('on_click', None)
+        self._click__handlers = []
+        super().__init__(**params)
+        if click_handler:
+            self.on_click(click_handler)
+
+    def _on__click(self, event: NodeEvent):
+        """
+        Method called when a node is clicked.
+        """
+
+    def on_click(
+        self, callback: Callable[[param.parameterized.Event], None | Awaitable[None]]
+    ) -> param.parameterized.Watcher:
+        """
+        Register a callback to be executed when the `Button` is clicked.
+
+        The callback is given an `Event` argument declaring the number of clicks
+
+        Example
+        -------
+
+        >>> button = pn.widgets.Button(name='Click me')
+        >>> def handle_click(event):
+        ...    print("I was clicked!")
+        >>> button.on_click(handle_click)
+
+        Arguments
+        ---------
+        callback:
+            The function to run on click events. Must accept a positional `Event` argument. Can
+            be a sync or async function
+
+        Returns
+        -------
+        watcher: param.Parameterized.Watcher
+          A `Watcher` that executes the callback when the button is clicked.
+        """
+        self._click__handlers.append(callback)
+
+    def _to_json(
+        self, id_, label, parent: str = None, children: Optional[list] = None,
+        icon: str = None, state: dict[str, bool] = {}, **kwargs
+    ):
+        if "selected" not in state:
+            state["selected"] = id_ in self.value
+        jsn = dict(id=id_, text=label, children=children or [], state=state, **kwargs)
+        if parent:
+            jsn["parent"] = parent
+        if icon:
+            jsn["icon"] = icon
+        else:
+            jsn["icon"] = "jstree-leaf"
+        return jsn
+
+    @classmethod
+    def _traverse(cls, node) -> list:
+        nodes = [node]
+        for subnode in node.get('children', []):
+            nodes += cls._traverse(subnode)
+        return nodes
+
+    def _reindex(self, reset=True):
+        if reset:
+            self._index = {}
+        for node in self._nodes:
+            for subnode in self._traverse(node):
+                self._index[subnode['id']] = subnode
```

Comment:
Potential problem with this, jstree doesn't require the initial node json data to include the id (https://www.jstree.com/docs/json/). If you don't include it, then jstree will provide it for you. Obviously the FileTree can assume an id because it is coded that way and it makes sense to make the id the unique file path, but a generic tree using the Tree class may not be constructed with an id and the nodes will only get an id after the first render with the jstree library. 

Since the FileTree uses this method to initialize the _index cache on on init, I wanted to mention that this won't work when work is completed to get the generic Tree class working again.